src/mdoc/builders/device-signed-builder.ts

```diff
@@ -0,0 +1,140 @@
+import type { MdocContext } from '../../context'
+import {
+  type CoseKey,
+  Header,
+  type MacAlgorithm,
+  ProtectedHeaders,
+  type SignatureAlgorithm,
+  UnprotectedHeaders,
+} from '../../cose'
+import { base64 } from '../../utils'
+import {
+  DeviceAuth,
+  DeviceMac,
+  DeviceNamespaces,
+  DeviceSignature,
+  DeviceSigned,
+  DeviceSignedItems,
+  type DocType,
+  type Namespace,
+  type SessionTranscript,
+} from '../models'
+import { DeviceAuthentication } from '../models/device-authentication'
+
+export class DeviceSignedBuilder {
+  private docType: DocType
+  private namespaces: DeviceNamespaces
+  private ctx: { cose: MdocContext['cose']; crypto: MdocContext['crypto'] }
+
+  public constructor(docType: DocType, ctx: { cose: MdocContext['cose']; crypto: MdocContext['crypto'] }) {
+    this.docType = docType
+    this.namespaces = new DeviceNamespaces({ deviceNamespaces: new Map() })
+    this.ctx = ctx
+  }
+
+  public addDeviceNamespace(namespace: Namespace, value: Record<string, unknown>) {
+    const deviceSignedItems =
+      this.namespaces.deviceNamespaces.get(namespace) ?? new DeviceSignedItems({ deviceSignedItems: new Map() })
+
+    for (const [k, v] of Object.entries(value)) {
+      deviceSignedItems.deviceSignedItems.set(k, v)
+    }
+
+    this.namespaces.deviceNamespaces.set(namespace, deviceSignedItems)
+
+    return this
+  }
+
+  public async sign(options: {
```

Comment:
not for now, but we probably want to revamp how this builder works, because each document usually has a different signing key. So providing a single key to sign doesn't make a lot of sense

Reply:
A device signed document can have multiple documents right?

Reply:
Oh wait that's device response